### PR TITLE
Do not hang on test failures of tests within `flutter_tools`

### LINKED ISSUE #141823: [flutter_tools] When running tool tests locally, test failures sometimes cause the runner to hang
Specifically, for tests with the `testUsingContext` wrapper, a test failure will cause the runner to hang until it's timed out after 15 minutes. Consider this minimal test:

```dart
import '../src/common.dart';
import '../src/context.dart';

void main() {
  testUsingContext('trivial test', () {
    fail('failure');
  });
}
```

### PR DESCRIPTION
**Pull request by matanlurey**
Fixes https://github.com/flutter/flutter/issues/141823

Before this change, when a test would fail, the terminal would hang (by default for 30s) until killed by the test runner.

Basically, [`runZonedGuarded`](https://api.flutter.dev/flutter/dart-async/runZonedGuarded.html) _does_ document (though not clearly) that a returned future should not be awaited:

```txt
The zone will always be an error-zone ([Zone.errorZone](https://api.flutter.dev/flutter/dart-async/Zone/errorZone.html)), so returning a future created inside the zone, and waiting for it outside of the zone, will risk the future not being seen to complete.
```

For example, you can see other places in Dart and Flutter that we circumvent that problem:

- https://github.com/flutter/flutter/blob/5987563e4aecb34fca446ea804943bb8d27d8fcd/packages/flutter_tools/test/general.shard/base/async_guard_test.dart#L279-L306
- https://github.com/dart-lang/dartdoc/blob/b04c9c127fea5f3fdf600aa205f50d81d1c779c5/lib/src/dartdoc.dart#L258-L264
- https://github.com/flutter/engine/blob/d1afda52d254f5f1faf79e51fe430d912d6db3ee/lib/web_ui/dev/browser_process.dart#L20-L22

I'm open to suggestions on how to test this :)

/cc @natebosch @jakemac53 @lrhn if you have any color commentary for us.

### CODE REVIEW

File: packages/flutter_tools/test/src/context.dart [hunk 1]
```diff
@@ -125,9 +125,21 @@ void testUsingContext(
           Analytics: () => NoOpAnalytics(),
         },
         body: () {
-          return runZonedGuarded<Future<dynamic>>(() {
+          // runZonedGuarded has a scary documenation that says:
```
Comment (christopherfujino) [inline packages/flutter_tools/test/src/context.dart:128]: sp. documentation

Comment (matanlurey) [reply packages/flutter_tools/test/src/context.dart]: Done.

File: packages/flutter_tools/test/src/context.dart [hunk 2]
```diff
@@ -125,9 +125,21 @@ void testUsingContext(
...
+          //
+          // "returning a future created inside the zone, and waiting for it
+          // outside of the zone, will risk the future not being seen to
+          // complete."
+          //
+          // for example, see our own test suite:
+          // https://github.com/flutter/flutter/blob/5987563e4aecb34fca446ea804943bb8d27d8fcd/packages/flutter_tools/test/general.shard/base/async_guard_test.dart#L279-L306
+          //
+          // so, we create a completer outside the zone, and complete it inside
+          // the zone in baiscally every terminal condition.
```
Comment (christopherfujino) [inline packages/flutter_tools/test/src/context.dart:138]: sp baiscally

I would suggest, but when I do the CLA bot gets confused

Comment (matanlurey) [reply packages/flutter_tools/test/src/context.dart]: Done.

**Review by christopherfujino — APPROVED** [review]
LGTM, pending unit test results of course

**Comment by Hixie** [comment]
What role is runZoneGuarded playing here?

**Comment by matanlurey** [comment]
> What role is runZoneGuarded playing here?

AFAICT, `runZonedGuarded` is the only way to catch 100% of asynchronous errors that occur as a side-effect.

For example:

```dart
// Will catch errors _returned_ by body, but not as a side-effect of body.
try {
  await body();
} catch (e) {
  print(e);
}
```

**Comment by christopherfujino** [comment]
> > What role is runZoneGuarded playing here?
> 
> AFAICT, `runZonedGuarded` is the only way to catch 100% of asynchronous errors that occur as a side-effect.
> 
> For example:
> 
> ```dart
> // Will catch errors _returned_ by body, but not as a side-effect of body.
> try {
>   await body();
> } catch (e) {
>   print(e);
> }
> ```

right. To put it another way, as far as I understand, `runZonedGuarded()` will create a new zone, run a `body` callback within that zone, and then catch all uncaught errors thrown from that zone.

File: packages/flutter_tools/test/src/context.dart [hunk 3]
```diff
@@ -145,14 +157,22 @@ void testUsingContext(
             } catch (error) { // ignore: avoid_catches_without_on_clauses
               _printBufferedErrors(context);
```
Comment (Hixie) [inline packages/flutter_tools/test/src/context.dart:158]: is this redundant with the one on line 169?

File: packages/flutter_tools/test/src/context.dart [hunk 4]
```diff
@@ -125,9 +125,21 @@ void testUsingContext(
...
+          //
+          // "returning a future created inside the zone, and waiting for it
+          // outside of the zone, will risk the future not being seen to
+          // complete."
+          //
+          // for example, see our own test suite:
+          // https://github.com/flutter/flutter/blob/5987563e4aecb34fca446ea804943bb8d27d8fcd/packages/flutter_tools/test/general.shard/base/async_guard_test.dart#L279-L306
+          //
+          // so, we create a completer outside the zone, and complete it inside
+          // the zone in basically every terminal condition.
```
Comment (Hixie) [inline packages/flutter_tools/test/src/context.dart:138]: I propose the following comment may be more helpful for folks in the future:

```
// To catch all errors thrown by the test, even uncaught async errors, we use a zone.
// Zones introduce their own event loop, so we do not await futures created inside
// the zone from outside the zone. Instead, we create a Completer outside the zone,
// and have the test complete it when the test ends (in success or failure), and we
// await that.
```

**Comment by jakemac53** [comment]
So this is super confusing and easy to get wrong, but it does actually make sense :).

The entire purpose of this API is to swallow errors, it does not allow them to escape the Zone. This includes async errors.

So, if there is an error attached to a Future which escapes the Zone (such as being returned from the `body` callback), that future will _never complete_, because it isn't allowed to complete with an error (outside the zone), and has no success value to complete with either. Instead, `onError` is called and that is it.

**Comment by jakemac53** [comment]
Fwiw I thought we added a different API which does essentially what your fix is, and has a return type of `R` instead of `R?`, with a required `onError` which has a return type of `R`. But I can't find it now. @natebosch am I making this up?

**Comment by Hixie** [comment]
@jakemac53 fwiw our general discontent around this is largely that Zones are basically entirely undocumented, so it's all guesswork when we try to use them. For example, https://github.com/dart-lang/sdk/issues/26455, https://github.com/dart-lang/sdk/issues/54177, https://github.com/dart-lang/sdk/issues/17667, https://github.com/dart-lang/sdk/issues/44669, https://github.com/dart-lang/sdk/issues/26454, ...

**Comment by jakemac53** [comment]
Sure, I have felt that pain many times before :). Even where there are docs I find them incredibly difficult to understand. I think the number of people who would self-proclaim to understand zones fully is 2, possibly just 1. I am not one of them :).

But, for this particular issue, I hope the simplified explanation I provided above helps illuminate what is actually happening here?

**Comment by Hixie** [comment]
I'm trying to encourage you to explain zones by submitting CLs to improve the API docs rather than just commenting on PRs. :-)

File: packages/flutter_tools/test/src/context.dart (hunk 3 above)
Comment (matanlurey) [reply packages/flutter_tools/test/src/context.dart:158]: No, these are errors caught within the executing future versus outside 😓

File: packages/flutter_tools/test/src/context.dart (hunk 4 above)
Comment (matanlurey) [reply packages/flutter_tools/test/src/context.dart:138]: Done.

File: packages/flutter_tools/test/src/context.dart (hunk 3 above)
Comment (natebosch) [reply packages/flutter_tools/test/src/context.dart:153]: This doesn't match my understanding - with the `rethrow` here I expect that `error` will also be passed to the `onError` callback argument, and `_printBufferedErrors(context);` will be called twice in that case.

Comment (matanlurey) [reply packages/flutter_tools/test/src/context.dart:153]: @natebosch Thanks! Is your suggestion cutting out this catch block or something else?

Comment (natebosch) [reply packages/flutter_tools/test/src/context.dart:153]: Yes, I think you can remove this `catch` and keep the `finally` to complete the `completer`.

**Comment by natebosch** [comment]
> Fwiw I thought we added a different API which does essentially what your fix is, and has a return type of `R` instead of `R?`, with a required `onError` which has a return type of `R`. But I can't find it now. @natebosch am I making this up?

I think you might be conflating another API. `Future.catchError` expects the error handler to return value with the same type as the Future but does not express this in the signature, and this caused some difficulty during the null safety migration. When we introduced `FutureExtensions.onError` we kept the same expectation but it's now checked statically.

We don't have any API that expects a value returned from a `Zone` error handler. There is no guarantee that this error happens at a time when it would be useful to complete a Future returned from the `runZonedGuarded` call - for example the following has both an early zone error and a later real value returned from the callback.

```dart
import 'dart:async';

void main() async {
  final value = await runZonedGuarded(() {
    Future.error('Some error'); // Not awaited - unhandled async error.
    return Future.delayed(const Duration(milliseconds: 500), () {
      return 'Some value';
    });
  }, (e, st) {
    print('Unhandled async error: $e\n$st');
  });
  print('Returned value: $value');
}
```

**Comment by natebosch** [comment]
@lrhn - WDYT about adding an API in `package:async` for this? The `Result` class might be a good place to stick it - that class is already dealing with treating errors as values. Maybe something like

```dart
  /// Runs [callback] in a zone and returns the result or the first
  /// unhandled error that occurs in the zone.
  ///
  /// If an unhandled asynchronous error occurs before the Future
  /// returned from [callback] has completed, the callback result
  /// will be ignored (if it ever completes), and the returned Future
  /// will complete with the same error.
  ///
  /// If there are multiple unhandled asynchronous errors, or if the
  /// Future returned by [callback] has an error after an unhandled
  /// asynchronous error, all but the first occurring error is ignored.
  ///
  /// This API may be useful to avoid Futures that never complete
  /// due to an error that cannot escape it's error zone.
  Future<T> Result.runZonedSmuggleErrors(Future<T> Function() callback);
```

Maybe we can also have a callback to handle those subsequent errors instead of strictly ignoring them.

**Comment by jakemac53** [comment]
Yeah I think that is really the API most people are expecting here.

**Pull request merged**